File: gui-client/world_step/type__entity/type__entity.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Ntype__entity {}
namespace XSD {
}
namespace XSD {
  public class type__entity : IEquatable<type__entity>, XSD.ILinkedNode  {

    public static string ClassTypeId = ".type__entity";
    public static string TagName = "type__entity";

    public string NodeName {get =>"type__entity";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<type__entity>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes
    private System.String _id;
    public System.String id { get => _id; set => _id = value; }
    private System.String? _entity_rule_ref;
    public System.String? entity_rule_ref { get => _entity_rule_ref; set => _entity_rule_ref = value; }

    //Children elements
    private XSD.Ntype__entity.text_map? _text_map = null;
    public XSD.Ntype__entity.text_map text_mapOrCreate
    {
      get
      {
        if(_text_map == null)
        {
          _text_map = new();
          _text_map.ParentNode = this;
          NotifyChange();
        }
        return _text_map;
      }
      set
      {
        _text_map = value;
        if(value != null)
        {
          value.ParentNode = this;
        }

      }
    }
    public XSD.Ntype__entity.text_map? text_map
    {
      get
      {
        return _text_map;
      }
      set
      {
        _text_map = value;
        if(value != null)
        {
          value.ParentNode = this;
        }
      }
    }

    private XSD.Ntype__entity.containers? _containers = null;
    public XSD.Ntype__entity.containers containersOrCreate
    {
      get
      {
        if(_containers == null)
        {
          _containers = new();
          _containers.ParentNode = this;
          NotifyChange();
        }
        return _containers;
      }
      set
      {
        _containers = value;
        if(value != null)
        {
          value.ParentNode = this;
        }

      }
    }
    public XSD.Ntype__entity.containers? containers
    {
      get
      {
        return _containers;
      }
      set
      {
        _containers = value;
        if(value != null)
        {
          value.ParentNode = this;
        }
      }
    }
    public type__entity()
    {
    }

    public type__entity(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public type__entity(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
      if(name == "id")
      {
        Set_id(value);
      }
      if(name == "entity_rule_ref")
      {
        Set_entity_rule_ref(value);
      }
    }

    public void SetChild(dynamic linkedNode)
    {
      if(linkedNode is XSD.Ntype__entity.text_map text_map)
      {
        this.text_map = text_map;
      }

      if(linkedNode is XSD.Ntype__entity.containers containers)
      {
        this.containers = containers;
      }

    }

    public void ClearChild(dynamic linkedNode)
    {
      if(linkedNode is XSD.Ntype__entity.text_map)
      {
        this.text_map = null;
      }

      if(linkedNode is XSD.Ntype__entity.containers)
      {
        this.containers = null;
      }

    }

    public Action OnSelfChange(Action<type__entity> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing type__entity");
      //Deserialize arguments
      if(rawNode.attributes.ContainsKey("id"))
      {
        var attribute_id = rawNode.attributes["id"];
        this.id = rawNode.attributes["id"];
      }
      if(rawNode.attributes.ContainsKey("entity_rule_ref"))
      {
        var attribute_entity_rule_ref = rawNode.attributes["entity_rule_ref"];
        this.entity_rule_ref = rawNode.attributes["entity_rule_ref"];
      }

      //Deserialize children
      text_map = rawNode.InitializeWithRawNode("text_map", text_map);

      containers = rawNode.InitializeWithRawNode("containers", containers);
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments
      if(this._id != null)
      {
        rawNode.attributes["id"] = this._id.ToString();
      }
      if(this._entity_rule_ref != null)
      {
        rawNode.attributes["entity_rule_ref"] = this._entity_rule_ref?.ToString();
      }

      //Serialize children
      if(text_map != null) {
        rawNode.children["text_map"] = new List<RawNode> { text_map.SerializeIntoRawNode() };
      }
      if(containers != null) {
        rawNode.children["containers"] = new List<RawNode> { containers.SerializeIntoRawNode() };
      }
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing type__entity");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_id()
    {
      return this.id;
    }
    public void Set_id(System.String value)
    {
      this.id = value;
      this.NotifyChange();
    }
    public System.String? Get_entity_rule_ref()
    {
      return this.entity_rule_ref;
    }
    public void Set_entity_rule_ref(System.String? value)
    {
      this.entity_rule_ref = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }
      if(xpath.StartsWith(XSD.Ntype__entity.text_map.TagName))
      {
        this.text_map ??= new XSD.Ntype__entity.text_map();
        var childXPath = xpath.Substring(XSD.Ntype__entity.text_map.TagName.Length + 3);
        this.text_map.DeserializeAtPath(childXPath, rawNode);
        return;
      }
      if(xpath.StartsWith(XSD.Ntype__entity.containers.TagName))
      {
        this.containers ??= new XSD.Ntype__entity.containers();
        var childXPath = xpath.Substring(XSD.Ntype__entity.containers.TagName.Length + 3);
        this.containers.DeserializeAtPath(childXPath, rawNode);
        return;
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      if(linkedNode is XSD.Ntype__entity.text_map casted_text_map) {
        return 0;
      }
      if(linkedNode is XSD.Ntype__entity.containers casted_containers) {
        return 0;
      }
      return null;
    }

    public bool IsValidChildType(ILinkedNode candidateChild) {
      return candidateChild is XSD.Ntype__entity.text_map
      || candidateChild is XSD.Ntype__entity.containers
      || false;
    }

    public bool Equals(type__entity? obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;

        var other = (type__entity)obj;
        return Equals(id, other.id) && Equals(entity_rule_ref, other.entity_rule_ref) && Equals(text_map, other.text_map) && Equals(containers, other.containers);
    }

    public override int GetHashCode()
    {
        var acc = 0;

        acc = HashCode.Combine(acc, id);
        acc = HashCode.Combine(acc, entity_rule_ref);
        acc = HashCode.Combine(acc, text_map);
        acc = HashCode.Combine(acc, containers);
        return acc;
    }
  }
}

File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/Type_entity/Type_entity.java
package ro.anud.xml_xsd.implementation.model.Type_entity;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.logging.LogScope.logScope;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class Type_entity implements  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static String nodeName = "type__entity";
    public static Type_entity fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        var instance = new Type_entity();
        if(Objects.nonNull(parent)) {
          instance.parentNode(parent);
        }
        instance.rawNode(rawNode);
        instance.deserialize(rawNode);
        return logger.logReturn(instance);
      }

    }
    public static Type_entity fromRawNode(RawNode rawNode) {
      try (var logger = logScope()) {
        var instance = fromRawNode(rawNode, null);
        return logger.logReturn(instance);
      }
    }
    public static Optional<Type_entity> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        try(var logger = logScope()) {
          return logger.logReturn(rawNode.map(o -> Type_entity.fromRawNode(o, parent)));
        }

    }
    public static List<Type_entity> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        List<Type_entity> returnList = Optional.ofNullable(rawNodeList)
            .orElse(List.of())
            .stream()
            .map(o -> Type_entity.fromRawNode(o, parent))
            .collect(Collectors.toList());
        return logger.logReturn(returnList);
      }
    }

    public String classTypeId() {
      return ".type__entity";
    }

    //Attributes

    private String id;
    @Builder.Default
    private Optional<String> entityRuleRef = Optional.empty();

    //Children elements
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap> textMap = Optional.empty();
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers> containers = Optional.empty();

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private RawNode rawNode = new RawNode();

    public RawNode rawNode() {
      return rawNode;
    }
    public void rawNode(RawNode rawNode) {
      this.rawNode = rawNode;
    }

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode() {
      return parentNode;
    }

    @Builder.Default
    private List<ro.anud.xml_xsd.implementation.util.ChangeCallback<Type_entity>> onChangeList = new ArrayList<>();
    @Builder.Default
    private List<ro.anud.xml_xsd.implementation.util.RemoveCallback<Type_entity>> onRemoveList = new ArrayList<>();

    public String nodeName() {
      return "type__entity";
    }
    public static Type_entity of() {
      return new Type_entity();
    }

    public void notifyChange(ro.anud.xml_xsd.implementation.util.LinkedNode object) {
      try (var logger = logScope()) {
        logger.log("Notify change for", this.buildPath());
        onChangeList.forEach(consumer -> consumer.onChange(object, this));
        parentNode.ifPresent(linkedNode -> linkedNode.notifyChange(object));
      }
    }

    public void notifyRemove(ro.anud.xml_xsd.implementation.util.LinkedNode object) {
      try (var logger = logScope()) {
        logger.log("Notify remove for", this.buildPath());
        onRemoveList.forEach(consumer -> consumer.onRemove(object, this));
        parentNode.ifPresent(linkedNode -> linkedNode.notifyRemove(object));
      }
    }

    public void parentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode.ifPresent(parent -> notifyRemove());
      this.parentNode = Optional.of(linkedNode);
      notifyChange();
    }

    public void removeChild(Object object) {
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap) {
          this.textMap = Optional.empty();
          notifyChange();
        }
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers) {
          this.containers = Optional.empty();
          notifyChange();
        }
    }

    public int buildIndexForChild(Object object) {
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap) {
          return 0;
        }
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers) {
          return 0;
        }
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(ro.anud.xml_xsd.implementation.util.ChangeCallback<Type_entity> callback) {
      try (var logger = logScope()) {
        onChangeList.add(callback);
        return logger.logReturn(() -> onChangeList.remove(callback));
      }
    }
    public Subscription onRemove(ro.anud.xml_xsd.implementation.util.RemoveCallback<Type_entity> callback) {
      try (var logger = logScope()) {
        onRemoveList.add(callback);
        return logger.logReturn(() -> onRemoveList.remove(callback));
      }
    }

    public void deserialize (RawNode rawNode) {
      try (var logger = logScope()) {
        this.rawNode = rawNode;
        var isDirty = false;
        try (var innerLogger = logScope("attributes")) {
          //Deserialize attributes
          innerLogger.log("id");
          var idValue = rawNode.getAttributeRequired("id");
          if(Objects.equals(this.id, idValue)) {
            isDirty = true;
          }
          this.id = idValue;
          innerLogger.log("entity_rule_ref");
          var entityRuleRefValue = rawNode.getAttribute("entity_rule_ref");
          if(Objects.equals(this.entityRuleRef, entityRuleRefValue)) {
            isDirty = true;
          }
          this.entityRuleRef = entityRuleRefValue;
        }
        try (var innerLogger = logScope("children")) {
          //Deserialize children
          this.textMap = ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap.fromRawNode(rawNode.getChildrenFirst("text_map"), this);
          this.containers = ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers.fromRawNode(rawNode.getChildrenFirst("containers"), this);
        }

        if(isDirty) {
          notifyChange();
        }
      } catch (Exception e) {
        throw new RuntimeException("Deserialization failed for: " + this.buildPath(), e);
      }

    }

    public RawNode serializeIntoRawNode()
    {
      try (var logger = logScope()) {
        rawNode.setTag("type__entity");
        try (var innerLogger = logScope("attributes")) {
          //Serialize attributes
          innerLogger.log("id");
          rawNode.setAttribute("id", this.id);
          innerLogger.log("entity_rule_ref");
          this.entityRuleRef.ifPresent(o -> rawNode.setAttribute("entity_rule_ref", o));
        }
        try (var innerLogger = logScope("children")) {

          //Serialize children
          innerLogger.log("text_map");
          rawNode.setChildren("text_map", textMap.stream().map(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap::serializeIntoRawNode).toList());
          innerLogger.log("containers");
          rawNode.setChildren("containers", containers.stream().map(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers::serializeIntoRawNode).toList());
          return rawNode;
        }
      }
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing type__entity");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }

    public String getId()
    {
      return this.id;
    }
    public Type_entity setId(String value)
    {
      this.id = value;
      notifyChange();
      return this;
    }
    public Optional<String> getEntityRuleRef()
    {
      return this.entityRuleRef;
    }
    public Type_entity setEntityRuleRef(Optional<String> value)
    {
      this.entityRuleRef = value;
      notifyChange();
      return this;
    }
    public Optional<ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap> getTextMap()
    {
      return this.textMap;
    }
    public ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap getTextMapOrDefault()
    {
      return this.textMap.orElseGet(() -> {
        var instance = new ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap();
        this.textMap = Optional.of(instance);
        instance.parentNode(this);
        return this.textMap.get();
      });
    }
    public java.util.stream.Stream<ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap> streamTextMapOrDefault()
    {
      return java.util.stream.Stream.of(getTextMapOrDefault());
    }
    public java.util.stream.Stream<ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap> streamTextMap()
    {
      return textMap.stream();
    }
    public Type_entity setTextMap(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap value)
    {
      this.textMap = Optional.ofNullable(value);
      value.parentNode(this);
      notifyChange();
      return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers> getContainers()
    {
      return this.containers;
    }
    public ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers getContainersOrDefault()
    {
      return this.containers.orElseGet(() -> {
        var instance = new ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers();
        this.containers = Optional.of(instance);
        instance.parentNode(this);
        return this.containers.get();
      });
    }
    public java.util.stream.Stream<ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers> streamContainersOrDefault()
    {
      return java.util.stream.Stream.of(getContainersOrDefault());
    }
    public java.util.stream.Stream<ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers> streamContainers()
    {
      return containers.stream();
    }
    public Type_entity setContainers(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers value)
    {
      this.containers = Optional.ofNullable(value);
      value.parentNode(this);
      notifyChange();
      return this;
    }

    public ro.anud.xml_xsd.implementation.util.LinkedNode deserializeAtPath(String xpath, RawNode rawNode) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }
        if(xpath.startsWith(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap.nodeName))
        {
          if(this.textMap.isEmpty()) {
            this.textMap = Optional.of(new ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap());
          }
          var childXPath = xpath.substring(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap.nodeName.length() + 3);
          return this.textMap.get().deserializeAtPath(childXPath, rawNode);
        }
        if(xpath.startsWith(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers.nodeName))
        {
          if(this.containers.isEmpty()) {
            this.containers = Optional.of(new ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers());
          }
          var childXPath = xpath.substring(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers.nodeName.length() + 3);
          return this.containers.get().deserializeAtPath(childXPath, rawNode);
        }

        deserialize(rawNode);
        return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> getNodeAtPath(String xpath) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }
        if(xpath.startsWith(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap.nodeName))
        {
          if(this.textMap.isEmpty()) {
            this.textMap = Optional.of(new ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap());
          }
          var childXPath = xpath.substring(ro.anud.xml_xsd.implementation.model.Type_entity.TextMap.TextMap.nodeName.length() + 3);
          return this.textMap.get().getNodeAtPath(childXPath);
        }
        if(xpath.startsWith(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers.nodeName))
        {
          if(this.containers.isEmpty()) {
            this.containers = Optional.of(new ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers());
          }
          var childXPath = xpath.substring(ro.anud.xml_xsd.implementation.model.Type_entity.Containers.Containers.nodeName.length() + 3);
          return this.containers.get().getNodeAtPath(childXPath);
        }
        return Optional.of(this);
    }
  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "attributes": {
          "metaType": "object",
          "value": {
            "id": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "entity_rule_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": true
            }
          }
        },
        "isSingle": true,
        "value": {
          "text_map": {
            "metaType": "object",
            "isSingle": true,
            "value": {
              "text": {
                "metaType": "object",
                "value": {},
                "isSingle": false,
                "isNullable": true,
                "attributes": {
                  "metaType": "object",
                  "value": {
                    "name": {
                      "metaType": "primitive",
                      "value": "xs:string",
                      "isNullable": false
                    },
                    "value": {
                      "metaType": "primitive",
                      "value": "xs:string",
                      "isNullable": false
                    }
                  }
                }
              }
            },
            "isNullable": true
          },
          "containers": {
            "metaType": "object",
            "isSingle": true,
            "value": {
              "container": {
                "metaType": "object",
                "attributes": {
                  "metaType": "object",
                  "value": {
                    "id": {
                      "metaType": "primitive",
                      "value": "xs:string",
                      "isNullable": false
                    },
                    "container_rule_ref": {
                      "metaType": "primitive",
                      "value": "xs:string",
                      "isNullable": false
                    }
                  }
                },
                "isSingle": false,
                "value": {
                  "entities": {
                    "metaType": "object",
                    "isSingle": true,
                    "value": {
                      "entity": {
                        "metaType": "object",
                        "value": {},
                        "isSingle": false,
                        "isNullable": true,
                        "attributes": {
                          "metaType": "object",
                          "value": {
                            "entity_id_ref": {
                              "metaType": "primitive",
                              "value": "xs:string",
                              "isNullable": false
                            }
                          },
                          "isNullable": false
                        }
                      }
                    },
                    "isNullable": true
                  }
                },
                "isNullable": true
              }
            },
            "isNullable": true
          }
        }
      },
      "name": "type__entity"
    }
  */